suseconnect/migration.go

```diff
@@ -466,11 +468,106 @@ func sortMigrationProducts(m connect.MigrationPath, installedIDs connect.StringS
 	})
 }
 
+// three-way comparison of editions (EDITION=VERSION[-RELEASE])
+// release part is ignored
+func compareEditions(left, right string) int {
+	// cut off (optional) release part
+	leftParts := strings.Split(left, "-")
+	rightParts := strings.Split(right, "-")
+	// split version into parts
+	leftParts = strings.Split(leftParts[0], ".")
+	rightParts = strings.Split(rightParts[0], ".")
+
+	// right-pad parts with zeros to match length
+	for len(leftParts) < len(rightParts) {
+		leftParts = append(leftParts, "0")
+	}
+	for len(rightParts) < len(leftParts) {
+		rightParts = append(rightParts, "0")
+	}
+
+	// lenghts are equal so we can use one index
+	for i := range leftParts {
+		var l, r int
+		var err error
+		// take and convert i-th part of left
+		if l, err = strconv.Atoi(leftParts[i]); err != nil {
+			panic(err)
+		}
+		// take and convert i-th part of right
+		if r, err = strconv.Atoi(rightParts[i]); err != nil {
+			panic(err)
+		}
+		if l < r {
+			return -1
+		}
+		if l > r {
+			return 1
+		}
+	}
+	return 0
+}
+
+func cleanupProductRepos(p connect.Product, force bool) error {
+	productPackages, err := connect.FindProductPackages(p.Name)
+	if err != nil {
+		return err
+	}
+	repos, err := connect.Repos()
+	if err != nil {
+		return err
+	}
+	for _, availableProduct := range productPackages {
+		// skip non-obsolete products
+		if compareEditions(availableProduct.Edition, p.Edition()) >= 0 {
+			continue
+		}
+		// filter out "(System Packages)" and already disabled repos
```

Comment:
what are "(System Packages)" ?

Reply:
These are packages which don't belong to any repo. They can be manually installed or be left over after some upgrade.

Reply:
The `zypper` man page says this:
```
       The set of installed packages on a system is sometimes denoted as
       repository @System or System Packages. In contrast to available
       repositories providing packages which can be installed, @System
       provides packages which can only be deleted. Installed packages which
       are not also provided by at least one of the available repositories are
       often denoted as being unwanted, orphaned or dropped.

Reply:
ah okay thanks